apps/src/lib/node/ledger/shell/finalize_block.rs

```diff
@@ -841,43 +845,71 @@ where
             pgf_inflation_amount.to_string_native()
         );
 
-        // Pgf steward inflation
-        let pgf_stewards_inflation_rate_key =
-            pgf_storage::get_steward_inflation_rate_key();
-        let pgf_stewards_inflation_rate = self
-            .read_storage_key::<Dec>(&pgf_stewards_inflation_rate_key)
-            .unwrap_or_default();
+        let mut pgf_fundings = pgf::get_payments(&self.wl_storage)?;
+        // we want to pay first the oldest fundings
+        pgf_fundings.sort_by(|a, b| a.id.cmp(&b.id));
 
-        let pgf_stewards_pd_rate =
-            pgf_stewards_inflation_rate / Dec::from(epochs_per_year);
-        let pgf_steward_inflation =
-            Dec::from(total_tokens) * pgf_stewards_pd_rate;
+        for funding in pgf_fundings {
+            if credit_tokens(
+                &mut self.wl_storage,
+                &staking_token,
+                &funding.detail.target,
+                funding.detail.amount,
+            )
+            .is_ok()
+            {
+                tracing::info!(
+                    "Minted {} tokens for {} project.",
+                    funding.detail.amount.to_string_native(),
+                    &funding.detail.target,
+                );
+            } else {
+                tracing::info!(
+                    "Failed Minting {} tokens for {} project.",
+                    funding.detail.amount.to_string_native(),
+                    &funding.detail.target,
+                );
+            }
+        }
 
-        let pgf_stewards_key = pgf_storage::get_stewards_key();
-        let pgf_stewards: BTreeSet<Address> =
-            self.read_storage_key(&pgf_stewards_key).unwrap_or_default();
+        // Pgf steward inflation
+        let stewards = pgf::get_stewards(&self.wl_storage)?;
 
-        let pgf_steward_reward = match pgf_stewards.len() {
+        let pgf_steward_reward = match stewards.len() {
             0 => Dec::zero(),
             _ => pgf_steward_inflation
-                .trunc_div(&Dec::from(pgf_stewards.len()))
+                .trunc_div(&Dec::from(stewards.len()))
                 .unwrap_or_default(),
         };
-        let pgf_steward_reward = token::Amount::from(pgf_steward_reward);
 
-        for steward in pgf_stewards {
-            credit_tokens(
-                &mut self.wl_storage,
-                &staking_token,
-                &steward,
-                pgf_steward_reward,
-            )?;
-            tracing::info!(
-                "Minting {} tokens for PGF Steward rewards distribution into \
-                 the steward address {}.",
-                pgf_steward_reward.to_string_native(),
-                steward,
-            );
+        for steward in stewards {
+            for (address, percentage) in steward.reward_distribution {
+                let pgf_steward_reward = pgf_steward_reward
+                    .checked_mul(&percentage)
+                    .unwrap_or_default();
```

Comment:
Should we log something in case of an overflow?

Reply:
this can never happen cause `percentage ` is between 0 and 1